File: mini-diffusion-go/diffusion/scheduler.go
package diffusion

import (
	"math"
	"math/rand"
)

// ScheduleType represents the noise schedule type.
type ScheduleType string

const (
	Linear    ScheduleType = "linear"
	Cosine    ScheduleType = "cosine"
	Quadratic ScheduleType = "quadratic"
)

// DiffusionConfig holds configuration for the diffusion process.
type DiffusionConfig struct {
	NumTimesteps int
	BetaStart    float64
	BetaEnd      float64
	Schedule     ScheduleType
}

// DefaultDiffusionConfig returns default configuration.
func DefaultDiffusionConfig() DiffusionConfig {
	return DiffusionConfig{
		NumTimesteps: 1000,
		BetaStart:    1e-4,
		BetaEnd:      0.02,
		Schedule:     Linear,
	}
}

// NoiseScheduler handles adding and removing noise at different timesteps.
type NoiseScheduler struct {
	Config                    DiffusionConfig
	Betas                     []float64
	Alphas                    []float64
	AlphasCumprod             []float64
	SqrtAlphasCumprod         []float64
	SqrtOneMinusAlphasCumprod []float64
	rng                       *rand.Rand
}

// NewNoiseScheduler creates a new noise scheduler.
func NewNoiseScheduler(config DiffusionConfig, rng *rand.Rand) *NoiseScheduler {
	betas := getBetas(config)

	alphas := make([]float64, len(betas))
	for i, b := range betas {
		alphas[i] = 1.0 - b
	}

	alphasCumprod := make([]float64, len(alphas))
	cumprod := 1.0
	for i, a := range alphas {
		cumprod *= a
		alphasCumprod[i] = cumprod
	}

	sqrtAlphasCumprod := make([]float64, len(alphasCumprod))
	sqrtOneMinusAlphasCumprod := make([]float64, len(alphasCumprod))
	for i, a := range alphasCumprod {
		sqrtAlphasCumprod[i] = math.Sqrt(a)
		sqrtOneMinusAlphasCumprod[i] = math.Sqrt(1.0 - a)
	}

	return &NoiseScheduler{
		Config:                    config,
		Betas:                     betas,
		Alphas:                    alphas,
		AlphasCumprod:             alphasCumprod,
		SqrtAlphasCumprod:         sqrtAlphasCumprod,
		SqrtOneMinusAlphasCumprod: sqrtOneMinusAlphasCumprod,
		rng:                       rng,
	}
}

// getBetas generates the beta schedule.
func getBetas(config DiffusionConfig) []float64 {
	t := config.NumTimesteps
	betas := make([]float64, t)

	switch config.Schedule {
	case Linear:
		for i := 0; i < t; i++ {
			betas[i] = config.BetaStart + (config.BetaEnd-config.BetaStart)*float64(i)/float64(t-1)
		}
	case Cosine:
		s := 0.008
		maxBeta := 0.999
		for i := 0; i < t; i++ {
			t1 := float64(i) / float64(t)
			t2 := float64(i+1) / float64(t)
			alphaBarT1 := math.Pow(math.Cos((t1+s)/(1.0+s)*math.Pi/2), 2)
			alphaBarT2 := math.Pow(math.Cos((t2+s)/(1.0+s)*math.Pi/2), 2)
			beta := 1.0 - alphaBarT2/alphaBarT1
			if beta > maxBeta {
				beta = maxBeta
			}
			betas[i] = beta
		}
	case Quadratic:
		for i := 0; i < t; i++ {
			frac := float64(i) / float64(t-1)
			betas[i] = config.BetaStart + (config.BetaEnd-config.BetaStart)*frac*frac
		}
	}

	return betas
}

// AddNoise adds noise to data at timestep t (forward diffusion).
// q(x_t | x_0) = N(x_t; sqrt(alpha_bar_t) * x_0, (1 - alpha_bar_t) * I)
// Returns: (noisy_data, noise)
func (s *NoiseScheduler) AddNoise(x0 *Tensor, t int) (*Tensor, *Tensor) {
	noise := Randn(x0.Shape, s.rng)
	noisy := s.AddNoiseWith(x0, noise, t)
	return noisy, noise
}

// AddNoiseWith adds specific noise to data at timestep t.
func (s *NoiseScheduler) AddNoiseWith(x0, noise *Tensor, t int) *Tensor {
	sqrtAlpha := s.SqrtAlphasCumprod[t]
	sqrtOneMinusAlpha := s.SqrtOneMinusAlphasCumprod[t]

	// x_t = sqrt(alpha_bar_t) * x_0 + sqrt(1 - alpha_bar_t) * noise
	return x0.Scale(sqrtAlpha).Add(noise.Scale(sqrtOneMinusAlpha))
}

// SampleTimesteps samples random timesteps for a training batch.
func (s *NoiseScheduler) SampleTimesteps(batchSize int) []int {
	timesteps := make([]int, batchSize)
	for i := range timesteps {
		timesteps[i] = s.rng.Intn(s.Config.NumTimesteps)
	}
	return timesteps
}

// Step computes a single denoising step.
// Given x_t and predicted noise, compute x_{t-1}
func (s *NoiseScheduler) Step(xT, predictedNoise *Tensor, t int) *Tensor {
	alphaT := s.Alphas[t]
	alphaBarT := s.AlphasCumprod[t]
	betaT := s.Betas[t]

	// Compute coefficient
	coef := betaT / math.Sqrt(1.0-alphaBarT)

	// Mean: (x_t - coef * predicted_noise) / sqrt(alpha_t)
	mean := xT.Sub(predictedNoise.Scale(coef)).Scale(1.0 / math.Sqrt(alphaT))

	// Add noise (except at final step)
	if t > 0 {
		sigma := math.Sqrt(betaT)
		noise := Randn(xT.Shape, s.rng)
		return mean.Add(noise.Scale(sigma))
	}

	return mean
}

// GetTimestepEmbedding creates sinusoidal timestep embeddings.
func GetTimestepEmbedding(timesteps []int, embDim int) *Tensor {
	batchSize := len(timesteps)
	halfDim := embDim / 2

	emb := Zeros([]int{batchSize, embDim})

	for b, t := range timesteps {
		for i := 0; i < halfDim; i++ {
			freq := math.Exp(-math.Log(10000.0) * float64(i) / float64(halfDim))
			arg := float64(t) * freq
			emb.Data.Set(b, i, math.Sin(arg))
			emb.Data.Set(b, i+halfDim, math.Cos(arg))
		}
	}

	return emb
}

File: mini-diffusion-python/mini_diffusion/scheduler.py
"""
Noise scheduler for diffusion models.

Implements linear, cosine, and quadratic beta schedules.
"""

import numpy as np
from enum import Enum
from typing import Optional

from .tensor import Tensor


class ScheduleType(Enum):
    """Type of noise schedule."""
    LINEAR = "linear"
    COSINE = "cosine"
    QUADRATIC = "quadratic"


class NoiseScheduler:
    """
    Noise scheduler for diffusion models.
    
    Implements the forward diffusion process q(x_t | x_0) and
    reverse process step predictions.
    """
    
    def __init__(
        self,
        num_timesteps: int = 1000,
        schedule_type: ScheduleType = ScheduleType.LINEAR,
        beta_start: float = 0.0001,
        beta_end: float = 0.02
    ):
        """
        Initialize noise scheduler.
        
        Args:
            num_timesteps: Total number of diffusion steps
            schedule_type: Type of beta schedule
            beta_start: Starting beta value
            beta_end: Ending beta value
        """
        self.num_timesteps = num_timesteps
        self.schedule_type = schedule_type
        
        # Generate beta schedule
        if schedule_type == ScheduleType.LINEAR:
            self.betas = np.linspace(beta_start, beta_end, num_timesteps)
        
        elif schedule_type == ScheduleType.COSINE:
            s = 0.008
            steps = np.arange(num_timesteps + 1)
            alphas_cumprod = np.cos((steps / num_timesteps + s) / (1 + s) * np.pi / 2) ** 2
            alphas_cumprod = alphas_cumprod / alphas_cumprod[0]
            self.betas = np.clip(1 - alphas_cumprod[1:] / alphas_cumprod[:-1], 0, 0.999)
        
        elif schedule_type == ScheduleType.QUADRATIC:
            self.betas = np.linspace(np.sqrt(beta_start), np.sqrt(beta_end), num_timesteps) ** 2
        
        else:
            raise ValueError(f"Unknown schedule type: {schedule_type}")
        
        # Compute alphas and cumulative products
        self.alphas = 1.0 - self.betas
        self.alphas_cumprod = np.cumprod(self.alphas)
        self.alphas_cumprod_prev = np.concatenate([[1.0], self.alphas_cumprod[:-1]])
        
        # Pre-compute useful quantities
        self.sqrt_alphas_cumprod = np.sqrt(self.alphas_cumprod)
        self.sqrt_one_minus_alphas_cumprod = np.sqrt(1.0 - self.alphas_cumprod)
        self.sqrt_recip_alphas_cumprod = np.sqrt(1.0 / self.alphas_cumprod)
        self.sqrt_recipm1_alphas_cumprod = np.sqrt(1.0 / self.alphas_cumprod - 1)
        
        # Posterior variance for q(x_{t-1} | x_t, x_0)
        self.posterior_variance = (
            self.betas * (1.0 - self.alphas_cumprod_prev) / (1.0 - self.alphas_cumprod)
        )
        self.posterior_log_variance_clipped = np.log(
            np.maximum(self.posterior_variance, 1e-20)
        )
        
        # Posterior mean coefficients
        self.posterior_mean_coef1 = (
            self.betas * np.sqrt(self.alphas_cumprod_prev) / (1.0 - self.alphas_cumprod)
        )
        self.posterior_mean_coef2 = (
            (1.0 - self.alphas_cumprod_prev) * np.sqrt(self.alphas) / (1.0 - self.alphas_cumprod)
        )
    
    @classmethod
    def linear(cls, num_timesteps: int = 1000) -> 'NoiseScheduler':
        """Create scheduler with linear beta schedule."""
        return cls(num_timesteps, ScheduleType.LINEAR)
    
    @classmethod
    def cosine(cls, num_timesteps: int = 1000) -> 'NoiseScheduler':
        """Create scheduler with cosine beta schedule."""
        return cls(num_timesteps, ScheduleType.COSINE)
    
    @classmethod
    def quadratic(cls, num_timesteps: int = 1000) -> 'NoiseScheduler':
        """Create scheduler with quadratic beta schedule."""
        return cls(num_timesteps, ScheduleType.QUADRATIC)
    
    def add_noise(
        self,
        sample: Tensor,
        noise: Tensor,
        timestep: int
    ) -> Tensor:
        """
        Add noise to sample at given timestep.
        
        q(x_t | x_0) = sqrt(alpha_cumprod_t) * x_0 + sqrt(1 - alpha_cumprod_t) * noise
        
        Args:
            sample: Original sample x_0
            noise: Gaussian noise
            timestep: Current timestep
            
        Returns:
            Noisy sample x_t
        """
        sqrt_alpha = self.sqrt_alphas_cumprod[timestep]
        sqrt_one_minus_alpha = self.sqrt_one_minus_alphas_cumprod[timestep]
        
        return sample * sqrt_alpha + noise * sqrt_one_minus_alpha
    
    def step(
        self,
        sample: Tensor,
        noise_pred: Tensor,
        timestep: int,
        rng: Optional[np.random.Generator] = None
    ) -> Tensor:
        """
        DDPM step: predict x_{t-1} from x_t and predicted noise.
        
        Args:
            sample: Current noisy sample x_t
            noise_pred: Predicted noise epsilon_theta(x_t, t)
            timestep: Current timestep
            rng: Random number generator
            
        Returns:
            Denoised sample x_{t-1}
        """
        if rng is None:
            rng = np.random.default_rng()
        
        # Predict x_0 from x_t and noise prediction
        x0_pred = (
            sample * self.sqrt_recip_alphas_cumprod[timestep]
            - noise_pred * self.sqrt_recipm1_alphas_cumprod[timestep]
        )
        
        # Compute posterior mean
        posterior_mean = (
            x0_pred * self.posterior_mean_coef1[timestep]
            + sample * self.posterior_mean_coef2[timestep]
        )
        
        # Add noise if not the last step
        if timestep > 0:
            noise = Tensor.randn(*sample.shape, rng=rng)
            variance = np.sqrt(self.posterior_variance[timestep])
            return posterior_mean + noise * variance
        
        return posterior_mean
    
    def step_ddim(
        self,
        sample: Tensor,
        noise_pred: Tensor,
        timestep: int,
        prev_timestep: int,
        eta: float = 0.0
    ) -> Tensor:
        """
        DDIM step: deterministic or stochastic sampling.
        
        Args:
            sample: Current noisy sample x_t
            noise_pred: Predicted noise epsilon_theta(x_t, t)
            timestep: Current timestep
            prev_timestep: Previous timestep
            eta: Noise level (0 for deterministic DDIM)
            
        Returns:
            Denoised sample x_{t-1}
        """
        alpha_cumprod = self.alphas_cumprod[timestep]
        alpha_cumprod_prev = self.alphas_cumprod[prev_timestep] if prev_timestep >= 0 else 1.0
        
        sqrt_alpha_cumprod = self.sqrt_alphas_cumprod[timestep]
        sqrt_one_minus_alpha_cumprod = self.sqrt_one_minus_alphas_cumprod[timestep]
        
        # Predict x_0
        x0_pred = (sample - sqrt_one_minus_alpha_cumprod * noise_pred) / sqrt_alpha_cumprod
        
        # DDIM formula
        sqrt_alpha_cumprod_prev = np.sqrt(alpha_cumprod_prev)
        sqrt_one_minus_alpha_cumprod_prev = np.sqrt(1.0 - alpha_cumprod_prev)
        
        return x0_pred * sqrt_alpha_cumprod_prev + noise_pred * sqrt_one_minus_alpha_cumprod_prev
    
    def get_timesteps(self, num_inference_steps: int) -> np.ndarray:
        """Get evenly spaced timesteps for inference."""
        step_ratio = self.num_timesteps // num_inference_steps
        timesteps = np.arange(0, num_inference_steps) * step_ratio
        timesteps = self.num_timesteps - 1 - timesteps
        return timesteps.astype(np.int64)
